rollbar/lib/_async.py

```diff
@@ -0,0 +1,219 @@
+    Asynchronously reports an arbitrary string message to Rollbar.
+
+    message: the string body of the message
+    level: level to report at. One of: 'critical', 'error', 'warning', 'info', 'debug'
+    request: the request object for the context of the message
+    extra_data: dictionary of params to include with the message. 'body' is reserved.
+    payload_data: param names to pass in the 'data' level of the payload; overrides defaults.
+    """
+
+    with async_handler():
+        try:
+            return await call_later(
+                _report_message(message, level, request, extra_data, payload_data)
+            )
+        except Exception as e:
+            log.exception('Exception while reporting message to Rollbar. %r', e)
+
+
+async def _report_exc_info(
+    exc_info=None, request=None, extra_data=None, payload_data=None, level=None, **kw
+):
+    return rollbar.report_exc_info(
+        exc_info, request, extra_data, payload_data, level, **kw
+    )
+
+
+async def _report_message(
+    message, level='error', request=None, extra_data=None, payload_data=None, **kw
+):
+    return rollbar.report_message(message, level, request, extra_data, payload_data)
+
+
+async def _post_api_httpx(path, payload_str, access_token=None):
+    headers = {'Content-Type': 'application/json'}
+    if access_token is not None:
+        headers['X-Rollbar-Access-Token'] = access_token
+    else:
+        headers['X-Rollbar-Access-Token'] = rollbar.SETTINGS.get('access_token')
+
+    proxy_cfg = {
+        'proxy': rollbar.SETTINGS.get('http_proxy'),
+        'proxy_user': rollbar.SETTINGS.get('http_proxy_user'),
+        'proxy_password': rollbar.SETTINGS.get('http_proxy_password'),
+    }
+    proxies = transport._get_proxy_cfg(proxy_cfg)
+
+    url = urljoin(rollbar.SETTINGS['endpoint'], path)
+    async with httpx.AsyncClient(
+        proxies=proxies, verify=rollbar.SETTINGS.get('verify_https', True)
+    ) as client:
+        resp = await client.post(
+            url,
+            data=payload_str,
+            headers=headers,
+            timeout=rollbar.SETTINGS.get('timeout', DEFAULT_TIMEOUT),
+        )
+
+    try:
+        return rollbar._parse_response(path, access_token, payload_str, resp)
+    except Exception as e:
+        log.exception('Exception while posting item %r', e)
+
+
+async def try_report(
+    exc_info=None, request=None, extra_data=None, payload_data=None, level=None, **kw
+):
+    current_handler = rollbar.SETTINGS.get('handler')
+    if not (current_handler in ALLOWED_HANDLERS or current_handler == 'default'):
+        raise RollbarAsyncError('No async handler set.')
+
+    if httpx is None:
+        raise RollbarAsyncError('HTTPX is required')
+
+    return await report_exc_info(
+        exc_info, request, extra_data, payload_data, level, **kw
+    )
+
+
+@contextlib.contextmanager
+def async_handler():
+    original_handler = rollbar.SETTINGS.get('handler')
+
+    if original_handler not in ALLOWED_HANDLERS:
```

Comment:
This doesn't seem safe at a quick glance, wouldn't the block running with this context manager be interleaved with other threads and / or coroutines, while `rollbar.SETTINGS` is modified?

Reply:
Great question!

This needs a bit more clarification, so let me put it in more structured answer.

### Short answer

Yes, this is not thread-safe and it's possible that `rollbar.SETTINGS` can be modified by other threads/coroutines while executing this context manager.

### Long answer

There are some important points that we need to take here under consideration.

#### Async and threads in Python

First of all, Python threads are very different from e.g. Java threads. It is unlikely that you will see mixed thread- and async- code together in Python. In fact, the `asyncio` module from the standard library was designed as a single-threaded framework. 

Here is the very first sentence from the `asyncio` documentation:
> This module provides infrastructure for writing **single-threaded** concurrent code using coroutines, multiplexing I/O access over sockets and other resources, running network clients and servers, and other related primitives.

Even though there are *related primitives* such as `asyncio.Lock`, they are still not thread safe:
> asyncio synchronization primitives are designed to be similar to those of the threading module with two important caveats: asyncio primitives are not thread-safe, therefore they should not be used for OS thread synchronization (use threading for that); (..)

More information: [PEP-3156](https://www.python.org/dev/peps/pep-3156/) and [asyncio docs](https://docs.python.org/3.9/library/asyncio-sync.html)

This is a theory. How is it handled in practice?

Most async web servers are run on top of an async server framework (e.g. `Uvicorn` or `Hypercorn`) that continue to be single-threaded. However, to increase performance and make a usage of all CPU cores, it's often combined with multiprocessing servers, such as `Gunicorn`. In practice, all event loops are run as *independent* processes without sharing memory.

Of course, some apps may be designed differently, combining async and threads for some reasons. In this case, the current implementation of the `async_handler` context manager **is not** thread safe.

I can add a comment on this to the handler docstrings.

#### Syncing async coroutines vs local context

The current design of the core of the SDK is really limited here. The app is built on global variables and public functions that assume to be executed as a singleton. Therefore, any runtime modifications to the SDK components should be considered as not thread-safe and performed with additional checks (most common problem:  notifier reinitialization).

Due to the current design of the core of the SDK (it's likely to be changed in the future), it's not possible to deliver usable (both async or thread) lock system too.

IMHO it's not a problem for threads as that's the nature of the current design and if someone wants to mix async and blocking reporting - well, the risk is known.

However, this issue may still occur for single-threaded applications based on coroutines.

Here is a short example that shows the problem of running two coroutines concurrently:

```python
import asyncio
import os

import rollbar
from rollbar.lib._async import report_message

rollbar.init(os.environ["RB_ACCESS_TOKEN"], 'async-test')

async def report(msg, handler):
    rollbar.SETTINGS['handler'] = handler
    await report_message(msg)

async def send_reports():
    await asyncio.gather(report('Report1', 'blocking'), report('Report2', 'httpx'))

if __name__ == '__main__':
    asyncio.run(send_reports())
```

In this case, the actual `Report1` handler is unknown as it could be `async` (due to the non-async handler and thanks to the `async_handler`ctx manager) or `httpx` (as `SETTINGS` are modified to the allowed `httpx` async handler on the second call).

As mentioned before, it's currently not possible to add async lock system for `async_handler`. In fact, even if it were possible, I'd rather avoid it.
 
One of the main advantage of using async code is performance. Adding safe-guards can affect actual performance by queuing reports to Rollbar API. Note that most of the core of the SDK is still synchronous and only called as awaitable block at a time (making it more async is on our roadmap). Locking handlers for each error would result in sequential reporting, one by one, almost from the beginning to the end of error processing.

However, I could use the local context feature. Instead of temporarily modifying `rollbar.SETTINGS`, I could give precedence to the local context handler in `send_payload()` which selects actual handler for dispatching. This should allow to maintain current performance using non-blocking operations and eliminate race conditions for the settings. The downside is that local context was introduced in Python 3.7+ and the current implementation of the async handler supports Python 3.6+.

### Possible solutions

#### Removal of the feature: `default` to `sync|async` handler

Let's start with the actual chain of function calls for ASGI, Starlette and FastAPI reporting:
https://github.com/rollbar/pyrollbar/blob/4cd225ee5297b11db81d161e4bce2ee66942cccf/rollbar/contrib/asgi/middleware.py#L24-L26

The `try_report()` coroutine has an additional safe-guard:
https://github.com/rollbar/pyrollbar/blob/4cd225ee5297b11db81d161e4bce2ee66942cccf/rollbar/lib/_async.py#L130-L142

Then `report_exc_info()` coroutine uses the `async_handler` context manager:
https://github.com/rollbar/pyrollbar/blob/4cd225ee5297b11db81d161e4bce2ee66942cccf/rollbar/lib/_async.py#L145-L159

As you can see, `RollbarAsyncError` is raised for anything other than `ALLOWED_HANDLERS` or non-`default` and then fails back to sync reporting. So, in theory, only `default` (or unset) handler can temporarily modify the settings via `handler_async` ctx manager.

By removing the `async_handler` ctx manager (and updating `try_report()` to `if not current_handler in ALLOWED_HANDLERS: raise RollbarAsyncError`), we lose the ability to "overwrite" the default/unset handler.

**Post-effect:** users must explicitly set the async handler. Using default/unset handler would raise `RollbarAsyncError`.

So the ASGI, Starlette and FastAPI integrations would actually use the `thread` handler by default.

Also, this basic example would actually raise `RollbarAsyncError` which can confuse users as well[1].

```python
import asyncio
import os

import rollbar
from rollbar.lib._async import report_message

rollbar.init(os.environ[RB_ACCESS_TOKEN'], 'async-test') # missing: handler="async" or "httpx"

async def send_report():
    await report_message('Hello world!')

if __name__ == '__main__':
    asyncio.run(send_report())
```

[1] Reporting coroutines are part of the private module `rollbar.lib._async`, so we can assume that users shouldn't use them directly anyway.

#### Adding support for local context

The only downside is that it is not compatible with Python 3.6. We would need to require Python 3.7+ for ASGI, Starlette, FastAPI and HTTPX integrations (ASGI is 3.5+, the rest is 3.6+).

#### Adding mixed solution

Use local context for Python 3.7+ and raise the exception / fails back to sync reporting when using Python 3.6 with not allowed async handler.

Due to inconsistency, I'm not a fan of this option. However, global access for request object (optional feature) is also based on local context variable and is supported only in Python 3.7+.

### Conclusions

Agree, this code is not thread/coro -safe and should be updated. However, at the current stage of the code base, it is not possible to deliver the desired approach.

I'm willing to add local context support for Python 3.7+ and a note in the context manager docs that it's not thread-safe. Would you approve this solution or would you vote to remove this feature altogether and use the `thread` handler by default until the codebase is redesigned?

Reply:
Excellent write-up!

Reply:
That seems like a decent compromise. It's understandable that with thread-based concurrency being what it is in Python (GIL and all...) it wasn't worth building a thread-safe implementation.

Now that async's in the picture and single-process concurrency is performant on Python, it would be worth progressively moving towards safety. But that's for another PR ;)

Reply:
@diegov Nah man. I don't like compromises ;)
I've just added 35da296 commit which adds local context support. I've also added support for 3.6 by backporting `contexvars` module via `aiocontextvars` ;-)

@ezarowny thanks!